core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java

```diff
@@ -191,7 +221,84 @@ public AtomicReference<Long> getWorkerId() {
 
   @Override
   public FileInfo getFileInfo(String fileId) throws IOException {
-    return new FileInfo();
+    alluxio.grpc.FileInfo fi;
+    String ufsFullPath = PathUtils.concatPath(mRootUFS, fileId);
+    String fn = new AlluxioURI(fileId).getName();
+
+    UfsFileStatus status = mUfsFileStatusCache.getIfPresent(ufsFullPath);
+    if (status == null) {
+      // The requested FileStatus is not present in memory cache.
+      // Let's try to query local persistent DoraMetaStore.
+      Optional<DoraMeta.FileStatus> fs = mMetaStore.getDoraMeta(ufsFullPath);
+      if (fs.isPresent()) {
+        // Found in persistent DoraMetaStore
+        fi = fs.get().getFileInfo();
+        String contentHash = UnderFileSystemUtils.approximateContentHash(fi.getLength(),
+            fi.getLastModificationTimeMs());
+        UfsFileStatus ufs = new UfsFileStatus(fi.getPath(), contentHash, fi.getLength(),
+            fi.getLastModificationTimeMs(),
+            fi.getOwner(), fi.getGroup(), (short) fi.getMode(), fi.getBlockSizeBytes());
+        mUfsFileStatusCache.put(ufsFullPath, ufs);
+      } else {
+        // This will load UfsFileStatus from UFS and put it in memory cache
+        try {
+          status = mUfsFileStatusCache.get(ufsFullPath);
+        } catch (ExecutionException e) {
+          Throwable throwable = e.getCause();
+          // this should be the exception thrown by ufs.getFileStatus which is IOException
+          if (throwable instanceof IOException) {
+            throw (IOException) throwable;
+          } else {
+            throw new InternalException("Unexpected exception when retrieving UFS file status",
+                throwable);
+          }
+        }
+        fi = buildFileInfoFromUfsFileStatus(status, fn, fileId, ufsFullPath);
+
+        // Add this to persistent DoraMetaStore.
+        long currentTimeMillis = System.currentTimeMillis();
+        mMetaStore.putDoraMeta(ufsFullPath,
+            DoraMeta.FileStatus.newBuilder().setFileInfo(fi).setTs(currentTimeMillis).build());
+      }
+    } else {
+      // Found in memory cache
+      fi = buildFileInfoFromUfsFileStatus(status, fn, fileId, ufsFullPath);
+    }
+    // because cache manager uses hashed ufs path as file ID
+    // TODO(bowen): we need a dedicated type for file IDs!
+    String cacheManagerFileId = new AlluxioURI(ufsFullPath).hash();
+
+    final long bytesInCache = mCacheManager.getUsage()
+        .flatMap(usage -> usage.partitionedBy(file(cacheManagerFileId)))
+        .map(CacheUsage::used).orElse(0L);
+    final long fileLength = fi.getLength();
+    final int cachedPercentage;
+    if (fileLength > 0) {
+      cachedPercentage = (int) (bytesInCache * 100 / fileLength);
+    } else {
+      cachedPercentage = 0;
+    }
```

Comment:
Performance concerns here.
What is the performance cost to do  `getUsage()`?

The Usage is only get updated in 'read()', and can we update the in-memory FileInfo cache there? So we don't have do `getUsage()` every time when we serve the getStatus() call.

Reply:
Very good point. I actually have thought about this a bit.

For the directory level usage stats, they are backed by a in-memory cached indicator, usually a `AtomicLong`, so getting it is fairly cheap.

For the scope and file level usage stats, the usage info is obtained by aggregating the pages those high level objects contains. For instance, file level usage is obtained by

https://github.com/Alluxio/alluxio/blob/61e9a78772d16f8378cf41d31c42fa1a1fdb50ba/core/client/fs/src/main/java/alluxio/client/file/cache/DefaultPageMetaStore.java#L245-L247

The costly part is summing up to several thousand `long`s to get the used space of a file.

We can definitely upfront calculate and update the used stat for each file on each `read` operation, so that obtaining the stats can be very cheap. But the downside is doing so needs even more memory, and if some files whose stats are never queried, we are essentially wasting the CPU time and memory.

These are the perf cost of the *current* implementation. If in the future we were to implement more complex queries like `used space of a file in all directories that are backed by a ramdisk`, and doing that requires a significant amount of computation, we may need to introduce some kind of upfront calculation when doing the reads.

Reply:
Let's keep to current design. We can improve the design and implementation if needed in future.